### DISCUSSION TITLE
2023-05-26 개발일지

### DISCUSSION THREAD

**Discussion by Craftcode17** [Blue-club/pintos1#1] [개발일지]
# TEAM 01 개발일지

### 프로젝트 본격시작 전 :
> - [x] 팀 규칙 정하기
> - [x] 집중 시간 : 하루에 2번 
( 13:30 : 학습할 것 정하고 학습 / 21:00 : 학습내용 공유, 질답 시간 )
> - [x] 각자의 작업사항 공유 : 22:00~ 22:15, 하루 마무리하는 시간
공유할 내용 : 오늘 한 것, 그 중 이슈사항, 내일 예정 작업
> - [x] 이론 공부방식
현재 구현과제에서 필요한 것을 그때그때 키워드 추출. 양이 많으면 분배하여 함께 공부한다
> - [x] 각자 아이디어 및 의견
당일 개발일지의 덧글에 임시 기록하기
> 
>

✏ 아이디어, 의견 : 당일 개발일지에 comment로 기록.   
✏ 팀플 원칙 : 최대한 간결하게 (불필요한 시간소모 유의)

---
### 내일의 할 일
- (공통) 과제 소스 파악, 작업구간 정하기
- (각자) 직접 수정해보기
---
### Alarm 과제에 당장 필요한 이론
이걸 러프하게 파악만 하고 과제 들어가자!

- 스레드 기초 ( 특히 스레드 상태)
- 스케줄링 (우선순위 스케줄링까지 x)
스레드 상태에 따라 실행/대기중/블록 상태 컨트롤
이 때에 쓰이는 관련 함수명과 그 역할
- 위 동작이 운영체제상에서 어떻게 구현되어 있는지
- 시스템 콜, 인터럽트, 등등    
   
cf. 당장 불필요 : 공유 자원 관련, 스케줄링 이론

### Alarm 작업 순서 :
1. 구동방식 파악
2. 수정할 곳 추측/확정
3. 수정 작업

**Comment by Craftcode17** [comment]
카톡은 기록남은걸 찾아보는게 번거로우니 팀작업 하는 걸 깃허브 기능으로 남겨봅시다
5조에서 기능 쓰는거 좋아보여서 시도함 :)

**Comment by Craftcode17** [comment]
1. 집중시간 제안 : 13:30 ~ (다같이 정하는 시간)
이유 : 밥먹은 후에 큰 집중이 필요한 작업을 바로 하기 힘들다. 팀 현황 파악하는 작업이 쉬우니 이 타이밍에 하면 좋을 것 같아서.

2. 일주일동안 세부 일정 정하기, 팀 계획에 맞춰 개인의 일주일 계획 세우자.
3. 이론키워드 3등분이 좀 어렵다. 아이디어 있으신 분? 
(동기화&동기화 보장 기법 / 스케줄링 / ..운영체제, 커널 등?? )
4. 이론 내용을 문서로 만드는것은 자율적으로(시간 너무 쓸 수 있으니 필수x). 
5. 단, 이론담당자에게 별도로 설명&질문 요청하는 것은 discussions에 남기자(또 찾아보게 될 것 같다)

**Comment by Riudiu** [comment]
1. 집중시간 정하기 - 13시 30분이나 14시 부터
2. 각자의 작업사항 공유는 밤에 퇴근하기 전 - 오늘한 것과 이슈사항 / 내일할 것 이렇게 작성하면 좋을 것 같습니다. 
3. 집중시간에 하루에 공부할 개념들을 정하고 각자 공부한 다음 21시쯤에 공부한거에 대해 얘기 나눠보는건 어떤가요? / 코드 작업할땐 각자 같은걸 작업하고 리뷰하고 제일 잘한 사람껄로 main에 머지 vs 각자 다른 부분 작업해서 리뷰하고 각자 작업한걸 main에 머지

**Reply by Craftcode17** [reply]
지우님 똑부러진다..!
작업사항 공유는 시간을 많이 쓰지 않으니 퇴근하기 전(대략  22시) 괜찮아 보입니다.

집중시간을 붙여서 하지 않고, 나눠서 한다 생각해도 좋을듯. 
- 점심직후 시간 : 공부할것 정하기
- 저녁시간 : 학습한 것 공유, 질답 등등

**Reply by Riudiu** [reply]
오 이렇게 하면 될거 같아요

**Comment by Craftcode17** [comment]
@carrotcookie @Riudiu

Alarm 과제에 필요할 듯한 이론(러프하게 파악만) 
1. 스레드 기초 ( 특히 스레드 상태)
2. 스케줄링 (우선순위 스케줄링까지 x)
- 스레드 상태에 따라 실행/대기중/블록 상태 컨트롤
- 이 때에 쓰이는 관련 함수명과 그 역할
3. 위 동작이 운영체제상에서 어떻게 구현되어 있는지
- 시스템 콜, 인터럽트, 등등

위 항목을 가볍게 배운 후에, 바로 alarm 과제 관련 소스를 보며
구동방식 파악
수정할 곳 추측/확정
수정 작업

**Comment by carrotcookie** [comment]
2023.05.26

context switching?
-> cpu/코어에서 실행 중이던 프로세스/스레드가 다른 프로세스/스레드로 교체되는 것. OS 커널에 의해 실행된다.

같은 프로세스의 스레드끼리 스위칭 - thread context switching
다른 프로세스끼리의 스위칭 - process context switching

thread context switching
-> t1, t2가 있다고 가정.
t1에서 t2로 넘어갈 때 t1의 cpu 상태 저장. t2의 cpu 상태를 로딩해온다.

process context switching
-> p1, p2가 있다고 가정.
p1에서 p2로 넘어갈 때 위와 동일하지만 추가로 MMU가 p1에서 p2의 메모리 영역을 가르키고 TLB를 비워줘야 함. -> 가상 메모리 영역이 달라지니까

synchronization가 왜 필요할까?

예를 들어 여러 프로세스/스레드가 동시에 같은 데이터를 조작한다고 가정하자. count라는 변수의 값을 +1 해주는 작업이라고 한다면.
count ++ 의 코드가 실행될 것이다. 따라서 컴퓨터는 count에 접근하여 값을 레지스터로 가져오고 레지스터에 값을 더하고 이를 다시 count에 반환하는 작업을 할것이다.
만약 T1(스레드)에서 count를 레지스터로 가져오고 나서 T2로 컨텍스트 스위칭이 일어나면 T1에서 1증가, T2에서 1증가하여 총 2증가를 기대했지만, 1만 증가하게 된다.

synchronization?
-> 여러 프로세스/스레드를 동시에 실행해도 공유 데이터의 일관성을 유지하는 것

race condition?
-> 여러 프로세스/스레드가 동시에 같은 데이터를 조작할 때 타이밍이나 접근 순서에 따라 결과가 달라질 수 있는 상황

critical section?
-> 공유 데이터의 일관성을 보장하기 위해 하나의 프로세스/스레드만 진입해서 실행 가능한 영역

synchronization을 위한 전략

어떻게 어느 프로세스/스레드에서 critical section에 접근 했다면 다른 프로세스/스레드에서는 접근하지 못하게 할까? -> lock을 사용하자

스레드가 실행되면서 lock을 걸고 critical section이 끝나면 lock을 해제한다.
-> 여러 스레드가 동시에 lock을 걸려 한다면?
-> cpu atomic 명령어를 통한 작업이기 때문에 실행 중간에 간섭받거나 중단되지 않고, 같은 메모리 영역에 대해 동시에 실행되지 않기때문에 괜찮.

spin lock?
-> lock을 가질 수 있을 때 까지 반복시도
-> 기다리는 동안 cpu 낭비하는 단점

mutex?
-> lock을 가질 수 있을 때까지 휴식

semaphore?
-> signal mechanism을 가진, 하나 이상의 프로세스/스레드가 critical section에 접근 가능하도록 하는 장치

**Comment by Riudiu** [comment]
https://riudiu.tistory.com/60
